### ISSUE TITLE
Add audio track to video, with custom initial frame number or timestamp

### ISSUE THREAD

**Issue by Kurisutian123** [bytedeco/javacv#1848]
I am generating an HLS stream in real-time.

I can generate image-only segment files with correct frameNumber / timestamp between them, and the HLS player can successfully play the segments, but as soon as I add the audio track, the HLS players can't play the segments anymore, it kind of seems like the audio tracks between the segments don't align or something, even though I use setFrameNumber.

Example:
```java
for (int i = 1; i <= totalFrames; i++) {
  recorder.setFrameNumber(frameCount);
  recorder.record(imageFrame, avutil.AV_PIX_FMT_RGB32);
  recorder.record(audioGrabber.grab());
}
```

If I comment out the record audio frame the whole HLS stream works fine, but as soon as audio is added the HLS players can't figure out how to play the stream anymore.

**Comment by Kurisutian123** [comment]
It's definitely some sort of discontinuity error.
Because if I add the HLS tag #EXT-X-DISCONTINUITY before and after each segment the stream starts working properly.
But I can't use #EXT-X-DISCONTINUITY in production reasons.

**Comment by Kurisutian123** [comment]
The discontinuity problem also occurs with this example below:

```java
for (int i = 1; i <= totalFrames; i++) {
  recorder.setFrameNumber(frameCount);
  recorder.record(imageFrame, avutil.AV_PIX_FMT_RGB32);
}

Frame audioFrame = new Frame();
ShortBuffer audioBuffer = ShortBuffer.allocate(44100 * 2 * VideoManager.VIDEO_LENGTH);
audioFrame.sampleRate = 44100;
audioFrame.audioChannels = 2;
audioFrame.samples = new ShortBuffer[] {audioBuffer};
recorder.record(audioFrame);
```
Yes I know, that is generating a silent audio track but even so it still causes the discontinuity error.
Remove the audioFrame recording block, and the discontinuity problem disappears.
But I need an audio track.

**Comment by saudet** [comment]
The length in time of an audio frame is not equal to a video frame. You'll need to synchronize (roughly) the streams.

**Comment by Kurisutian123** [comment]
I can't make any sense of any of this.
I've been reading example video files from the internet and extracting information to compare it to the files I generate.

```
Count | Type | Frame Number | Timestamp
1. (other) 0 - 0
1. (audio) 0 - 62017000
1. (audio) 0 - 62038333
1. (audio) 0 - 62059666
1. (audio) 0 - 62081000
1. (audio) 0 - 62102333
1. (audio) 0 - 62123666
1. (audio) 0 - 62145000
1. (audio) 0 - 62166333
1. (audio) 0 - 62187666
1. (image) 3721 - 62032000
2. (audio) 3721 - 62209000
2. (image) 3722 - 62049000
3. (audio) 3722 - 62230333
3. (image) 3723 - 62066000
4. (image) 3724 - 62082000
5. (audio) 3724 - 62251666
5. (image) 3725 - 62099000
6. (audio) 3725 - 62273000
6. (image) 3726 - 62116000
7. (audio) 3726 - 62294333
7. (image) 3727 - 62132000
8. (audio) 3727 - 62315666
8. (image) 3728 - 62149000
9. (image) 3729 - 62166000
10. (audio) 3729 - 62337000
10. (image) 3730 - 62183000
```
The image frames and audio frames of all the samples I've found on the internet are not really synced at all.
Sometimes there are multiple audio frames before a single image frame, sometimes there are no audio frames in front of image frames.
Doesn't make much sense.
Source: A Twitch VOD, 60 frames per second.

**Comment by Kurisutian123** [comment]
How is the timestamp calculated?
Should I set frame number and timestamp?
I see the frame number is only increasing when there is an image frame BUT the timestamp is increasing even when there is no image frame.

**Comment by Kurisutian123** [comment]
The image timestamp between different frames are 17000 which must be 1_000_000L / 60 (frame rate).
But how is the audio timestamp calculated and does setTimestamp even affect audio frames?

**Comment by saudet** [comment]
>The image frames and audio frames of all the samples I've found on the internet are not really synced at all.
Sometimes there are multiple audio frames before a single image frame, sometimes there are no audio frames in front of image frames.

As long as they are roughly synced, it will work.

> But how is the audio timestamp calculated and does setTimestamp even affect audio frames?

We can't set the timestamp for the audio frames, no. The size of the audio frames that you get depends on the codec.

**Comment by Kurisutian123** [comment]
Is there no way to do what I need to do then?
Not only do I need to do this, I seem to have figure out roughly how to get one clip going.
But I also need to get video clips, and change their initial timestamp.
To create fragments on the fly.
Like for instance I might need this one clip to start with the frame number 25000 and the timestamp 1_000_000_000, the image frames successfully work, but the audio track breaks the whole thing and video players refuse to play it, because the timestamp of the audio frames are unchanged.

**Comment by saudet** [comment]
The timestamps of the audio frames are just determined by their amount, so when you need less of them, call record() less often, or more often, if you need more of them.

**Comment by Kurisutian123** [comment]
I kind of roughly just solved the problem, by mutating every single ``frame.timestamp`` value including audio frames.
Thanks for your time!

**Comment by Kurisutian123** [comment]
I spent ages trying to figure this out.

First I started with ``recorder.setFrameNumber()``, then with ``recorder.setTimestamp()`` and then as a last resort I tried changing ``frame.timestamp`` directly and it kind of worked with the occasional error.

All I wanted to do was change the initial timestamp of the video so I could chain other MPEG-TS segments before it.
To change **_"initial timestamp"_**, and then it hit me, I searched on google for video offset and I found it.

The actual solution that I was looking for:
``recorder.setOption("output_ts_offset", String.valueOf(initialTimestamp));``

It's safe to say I'm new to the whole video encoding subject.
